q2_moshpit/busco/types/_format.py

```diff
@@ -48,3 +48,101 @@ def _validate_(self, level):
     'BUSCOResultsDirectoryFormat', 'busco_results.tsv',
     BUSCOResultsFormat
 )
+
+
+class BuscoGenericTextFileFmt(model.TextFileFormat):
+    def _validate_(self, level):
+        pass
+
+
+class BuscoGenericBinaryFileFmt(model.BinaryFileFormat):
+    def _validate_(self, level):
+        pass
+
+
+class BuscoDatabaseDirFmt(model.DirectoryFormat):
+    # File collections for text files
+    (
+        ancestral,
+        dataset,
+        lengths_cutoff,
+        scores_cutoff,
+        links_to_ODB,
+        ancestral_variants,
+        ogs_id,
+        species,
+        hmms,
+        refseq_db_md5
+    ) = [
+            model.FileCollection(
+                rf"busco_downloads\/lineages\/.+\/{pattern}",
+                format=BuscoGenericTextFileFmt
+            )
+            for pattern in [
+                r'ancestral$',
+                r'dataset\.cfg$',
+                r'lengths_cutoff$',
+                r'scores_cutoff$',
+                r'links_to_ODB.+\.txt$',
+                r'ancestral_variants$',
+                r'info\/ogs\.id\.info$',
+                r'info\/species\.info$',
+                r'hmms\/.+\.hmm$',
+                r'refseq_db\.faa\.gz\.md5'
+            ]
+        ]
+
+    # Placement_files. Optional because they are not in virus DB
+    (
+        list_of_reference_markers,
+        mapping_taxid_lineage,
+        mapping_taxids_busco_dataset_name,
+        tree,
+        tree_metadata,
+    ) = [
+            model.FileCollection(
+                rf"busco_downloads\/placement_files\/{pattern}",
+                format=BuscoGenericTextFileFmt,
+                optional=True
+            )
+            for pattern in [
+                r'list_of_reference_markers\..+\.txt$',
+                r'mapping_taxid-lineage\..+\.txt$',
+                r'mapping_taxids-busco_dataset_name\..+\.txt$',
+                r'tree\..+\.nwk$',
+                r'tree_metadata\..+\.txt$',
+            ]
+        ]
+
+    # Others
+    supermatrix_aln = model.FileCollection(
+        r'busco_downloads\/placement_files\/supermatrix\.aln\..+\.faa$',
+        format=AlignedProteinFASTAFormat,
+        optional=True
+    )
+    prfls = model.FileCollection(
+        r'busco_downloads\/lineages\/.+\/prfl\/.+\.prfl$',
+        format=BuscoGenericTextFileFmt,
+        optional=True
+    )
+    version_file = model.File(
+        'busco_downloads/file_versions.tsv', format=BuscoGenericTextFileFmt
+    )
+    refseq_db = model.FileCollection(
+        r'busco_downloads\/lineages\/.+refseq_db\.faa\.gz',
+        format=BuscoGenericBinaryFileFmt
+    )
+
+    def _path_maker(self, name):
+        return str(name)
```

Comment:
Why do we need to have this function?

Reply:
So essentially every file collection defined in the class needs a `<variable_name>_path_maker` function. Here what I do is define a generic function (`_path_maker`) and then upon initialization of an instance of the class (`__init__` method), I instantiate a path maker function for each of the file collection variables. The alternative to this would be defining a function for each variable, e.g. 

```python
def refseq_db_path_maker(self, name):
        return str(name)

def version_file_path_maker(self, name):
        return str(name)

...
```

I found this would be too redundant.

Reply:
ok yes, that makes sense.